### PR TITLE
ENH: `stats.quantile`: support `nan_policy='omit'` with JAX backend

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-20544

#### What does this implement/fix?
This allows `stats.quantile` to support `nan_policy='omit'` when JAX is the backend.

#### Additional information
~~Still thinking about lines 121-126. They're probably fine; just wondering why there was `elif` before.~~ There was a reason, but switching the order (as I had done) and making both `if` is even better. Added a test case that would fail in main, would fail in `main` for _two_ reasons if both statements were `if`, and passes in this PR.

~~Also, I should check whether this allows us to remove any other JAX skips.~~ I don't see any.

#### AI Generation Disclosure
No AI

### CODE REVIEW

File: scipy/stats/tests/test_quantile.py [hunk 1]
```diff
@@ -365,6 +363,22 @@ def test_unity_weights(self, method, shape, xp):
         ref = stats.quantile(x, p, method=method)
         xp_assert_close(res, ref)
 
+    def test_all_nan_harrell_davis_gh24707(self, xp):
```
Comment (mdhaber) [inline scipy/stats/tests/test_quantile.py:368]:
```suggestion
    @skip_xp_backends(cpu_only=True, reason="PyTorch doesn't have `betainc`.",
                      exceptions=['cupy', 'jax.numpy'])
    def test_all_nan_harrell_davis_gh24707(self, xp):
```

Comment (mdhaber) [reply scipy/stats/tests/test_quantile.py:368]: This looks like the only real failure, and it's no surprise - we skip torch + GPU + Harrell-Davis above. The rest are in main.

**Comment by mdhaber** [comment]
Failures are unrelated. I think this is GTG.

File: scipy/stats/_quantile.py [hunk 2]
```diff
@@ -104,29 +104,26 @@ def _quantile_iv(x, p, method, axis, nan_policy, keepdims, weights):
...
-    # JIX, so we can come back to this later.
-    if not is_lazy_array(y) and contains_nans:
+    if is_lazy_array(y) or contains_nans:
         nans = xp.isnan(y)
 
         # Note that if length along `axis` were 0 to begin with,
         # it is now length 1 and filled with NaNs.
         if nan_policy == 'propagate':
-            nan_out = xp.any(nans, axis=-1)
+            nan_out = xp.any(nans, axis=-1, keepdims=True)
```
Comment (mdhaber) [inline scipy/stats/_quantile.py:113]: `nans` is a boolean mask of all the NaNs in `y`. We used to be able to let `any` reduce away the last axis here, then later use the reduced array to set *entire* slices of `y` to NaN. That is not array API compatible. Now, we keep the dimension and broadcast it to the same shape of `y` later. *That* is used to set all elements of slices containing NaNs to NaN.

File: scipy/stats/_quantile.py [hunk 3]
```diff
@@ -104,29 +104,26 @@ def _quantile_iv(x, p, method, axis, nan_policy, keepdims, weights):
...
         else:  # 'omit'
             n_int = n - xp.count_nonzero(nans, axis=-1, keepdims=True)
             n = xp.astype(n_int, dtype)
             # NaNs are produced only if slice is empty after removing NaNs
-            nan_out = xp.any(n == 0, axis=-1)
+            nan_out = n == 0
```
Comment (mdhaber) [inline scipy/stats/_quantile.py:118]: There was no reason for `any` before; `n` has length one along the last axis.

File: scipy/stats/_quantile.py [hunk 4]
```diff
@@ -104,29 +104,26 @@ def _quantile_iv(x, p, method, axis, nan_policy, keepdims, weights):
...
-        if xp.any(nan_out):
-            y = xp.asarray(y, copy=True)  # ensure writable
-            y = xpx.at(y, nan_out).set(xp.nan)
-        elif xp.any(nans) and method == 'harrell-davis':
+        if (is_lazy_array(nans) or xp.any(nans)) and (method == 'harrell-davis'):
             y = xp.asarray(y, copy=True)  # ensure writable
             y = xpx.at(y, nans).set(0)  # any non-nan will prevent NaN from propagating
+        if is_lazy_array(nan_out) or xp.any(nan_out):
+            y = xp.asarray(y, copy=True)  # ensure writable
+            y = xpx.at(y, xp.broadcast_to(nan_out, y.shape)).set(xp.nan)
```
Comment (mdhaber) [inline scipy/stats/_quantile.py:126]: The logic here is the same but the order is swapped. The high-level understanding is that `nan_out` has the final say over whether the result should be NaN or not; the `nan` thing for "harrell-davis" is just to avoid the result becoming NaN accidentally, but the next part can override it.

The new `test_all_nan_harrell_davis_gh24707` is really testing this part, which was wrong before but works now.

I'm going to have to revisit this again in gh-23949 anyway. It can probably be simplified further, but I'd rather not polish it here to have to edit it again when `ogive` (or whatever it ends up being called) needs to use the same code.

**Review by steppi — APPROVED** [review]
Thanks @mdhaber. It took me a little longer than expected to understand what's going on, but it makes sense to me now. `nan_out` only if the slice contains only NaNs is correct behavior and I understand the branch that is doing that. I can see that Harrel-Davis is the only method that will touch `y` in such a way that NaNs could propagate, and can see that the choice of placeholder value doesn't impact the output due to how `w` is calculated here:

https://github.com/mdhaber/scipy/blob/ae003954f9c21bf5315fce300fef39d32ea4e5ab/scipy/stats/_quantile.py#L448-L452

This could perhaps benefit from comments explaining more of the logic, but I don't think that's necessary to do here before merging.

**Comment by mdhaber** [comment]
Thanks! Can probably add more detailed comments in the `ogive` pr
